### PR TITLE
Support hyperlinks in supported terminals

### LINKED ISSUE #35: Add support for hyperlink
[Issuehunt badges](https://issuehunt.io/r/chalk/wrap-ansi/issues/35)

Add support for hyperlinks

image

You can see that this package correctly split words with background color, but hyperlink remain untouchable. I add indent, and I want to split hyperlink also.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $80.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
### Submitted pull Requests
- [#37 Tweaking to support hyperlinks in supporting terminals](https://issuehunt.io/r/chalk/wrap-ansi/pull/37)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/chalk/wrap-ansi/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by nicholaschiasson**
Fixes #35

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#35: Add support for hyperlink](https://issuehunt.io/repos/41053676/issues/35)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by nicholaschiasson** [comment]
#35

**Comment by sindresorhus** [comment]
// @stroncium @TiagoDanin Would you be able to help review?

File: package.json [hunk 1]
```diff
@@ -48,7 +48,8 @@
 	"dependencies": {
 		"ansi-styles": "^4.0.0",
 		"string-width": "^4.1.0",
-		"strip-ansi": "^5.0.0"
+		"strip-ansi": "^5.0.0",
+		"terminal-link": "^2.0.0"
```
Comment (sindresorhus) [inline package.json]: This should be a dev dependency.

File: index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 'use strict';
+const ansiStyles = require('ansi-styles');
 const stringWidth = require('string-width');
 const stripAnsi = require('strip-ansi');
-const ansiStyles = require('ansi-styles');
```
Comment (sindresorhus) [inline index.js]: Don't do unrelated changes.

File: index.js [hunk 3]
```diff
@@ -36,12 +40,20 @@ const wrapWord = (rows, word, columns) => {
...
 		if (ESCAPES.has(character)) {
 			isInsideEscape = true;
-		} else if (isInsideEscape && character === 'm') {
-			isInsideEscape = false;
-			continue;
+			isInsideLinkEscape = word.slice(index).startsWith(`${character}${ANSI_ESCAPE_LINK}`);
 		}
 
 		if (isInsideEscape) {
+			/* istanbul ignore if: cannot enter on terminals not supporting hyperlinks */
```
Comment (stroncium) [inline index.js]: The code doesn't take into account terminal capabilities, so why is this ignored for coverage?

File: index.js [hunk 4]
```diff
@@ -36,12 +40,20 @@ const wrapWord = (rows, word, columns) => {
...
+			if (isInsideLinkEscape) {
+				if (character === ANSI_ESCAPE_BELL) {
+					isInsideEscape = false;
+					isInsideLinkEscape = false;
+				}
+			} else if (character === 'm') {
```
Comment (stroncium) [inline index.js]: `'m'` should be defined as constant.

File: index.js [hunk 5]
```diff
@@ -36,12 +40,20 @@ const wrapWord = (rows, word, columns) => {
 
 		if (ESCAPES.has(character)) {
 			isInsideEscape = true;
-		} else if (isInsideEscape && character === 'm') {
-			isInsideEscape = false;
-			continue;
+			isInsideLinkEscape = word.slice(index).startsWith(`${character}${ANSI_ESCAPE_LINK}`);
```
Comment (stroncium) [inline index.js]: This can be done without rendering a string on every escape.

File: test.js [hunk 6]
```diff
@@ -148,3 +149,21 @@ test('#27, does not remove spaces in line with ansi escapes when no trimming', t
 	t.is(wrapAnsi(chalk.bgGreen(`  ${chalk.black('OK')} `), 100, {trim: false}), chalk.bgGreen(`  ${chalk.black('OK')} `));
 	t.is(wrapAnsi(chalk.bgGreen(' hello '), 10, {hard: true, trim: false}), chalk.bgGreen(' hello '));
 });
+
+test('#35, wraps hyperlinks, preserving clickability in supporting terminals', t => {
+	const link = 'https://testlogin:testpassword@areallylongurlthatneedstogetwrapped.com/with_some/path?andparams=bogus&more=evenmorebogus#evensomehashparams=notdoinganything';
```
Comment (stroncium) [inline test.js]: It would be nice if the text in test cases was much shorter. At the moment it's hard to even figure out what is happening.

Comment (stroncium) [reply test.js]: UPD: In the end, I failed to understand what is happening in the end so didn't review the test cases.

File: index.js (hunk 4 above)
Comment (nicholaschiasson) [reply index.js]: Not really part of my change but sure I'll fix it.

File: test.js (hunk 6 above)
Comment (nicholaschiasson) [reply test.js]: Hopefully I clarified things a bit. Let me know if it's still incomprehensible though.

Comment (stroncium) [reply test.js]: Probably could use splitting the lines in expected too. But after some editing i can see the logic now.

Anyways, I tried this in my terminal(`termite`) and on a couple others(from supported list at https://github.com/sindresorhus/terminal-link, I tried `guake` and `terminology`), input works, output not clickable on lines beyond first, not displayed on lines beyond first, a click on first line results in opening incorrect link.

Comment (stroncium) [reply test.js]: Also, after printing this, all terminals entered state where nothing that is printed is displayed at all(until reset).

**Comment by sindresorhus** [comment]
@nicholaschiasson Ping :)

**Comment by nicholaschiasson** [comment]
Just updating, I plan to resolve all of the open discussions here soon! Sorry for the wait 😄

File: test.js (hunk 6 above)
Comment (nicholaschiasson) [reply test.js:154]: After having tested in a number of terminal emulators, it seems that many don't even support ansi links. I did a simple test using terminal-link in DomTerm, Terminology, and iTerm2 (which are all stated to support it), as well as Hyper and Termite (which are noted to have feature requests sent) and Terminus. Only iTerm2 exhibited the correct behaviour. I'm not why DomTerm and Terminology didn't work... I wasn't able to build guake to test.

In any case, I think I know the reason printing the output breaks the terminal display. I will work on this.

**Comment by nicholaschiasson** [comment]
Addressed all discussions, but apparently support for node 8 is broken. I will look into this soon.

**Comment by sindresorhus** [comment]
Support for Node.js 8 doesn’t matter. I plan to target Node.js 10 after this PR.

**Comment by nicholaschiasson** [comment]
> Support for Node.js 8 doesn’t matter. I plan to target Node.js 10 after this PR.

Ah whoops, I fixed it nonetheless. Up to you which style is preferable, named regex capture groups or not, but in any case I think this is ready again for review.

**Comment by sindresorhus** [comment]
Can you revert https://github.com/chalk/wrap-ansi/pull/37/commits/2b6a0b4050412cc074214650c2ce73e219316b4c?

**Comment by sindresorhus** [comment]
Thank you :)
